src/index.js

```diff
@@ -274,26 +292,48 @@ const aws = require("aws-sdk");
 aws.config.region = "${this.options.region}";
 const lambda = new aws.Lambda();
 const functionNames = ${JSON.stringify(functionNames)};
+const functionConcurrency = ${JSON.stringify(fullFunctionConcurrency)};
 module.exports.warmUp = async (event, context, callback) => {
   console.log("Warm Up Start");
   const invokes = await Promise.all(functionNames.map(async (functionName) => {
-    const params = {
-      ClientContext: "${Buffer.from(`{"custom":${this.warmup.source}}`).toString('base64')}",
-      FunctionName: functionName,
-      InvocationType: "RequestResponse",
-      LogType: "None",
-      Qualifier: process.env.SERVERLESS_ALIAS || "$LATEST",
-      Payload: '${this.warmup.source}'
-    };
-
-    try {
-      const data = await lambda.invoke(params).promise();
-      console.log(\`Warm Up Invoke Success: \${functionName}\`, data);
-      return true;
-    } catch (e) {
-      console.log(\`Warm Up Invoke Error: \${functionName}\`, e);
-      return false;
+    let concurrency = functionConcurrency[functionName] > 0 ? functionConcurrency[functionName] : 1;
+    let source = ${this.warmup.source};
+    let promises = [];
+    
+    console.log(\`Warming up function: \${functionName} with concurrency: \${concurrency}\`);
+    
+    for (let x = 0; x < concurrency; x++) {
+      source.concurrencyIndex = x;
+      const params = {
+        ClientContext: Buffer.from(JSON.stringify({"custom":source})).toString('base64'),
+        FunctionName: functionName,
+        InvocationType: "RequestResponse",
+        LogType: "None",
+        Qualifier: process.env.SERVERLESS_ALIAS || "$LATEST",
+        Payload: JSON.stringify(source)
+      };
+      
+      promises.push(lambda.invoke(params).promise());
     }
+    
+    let success = true;
+    await Promise.all(promises).then(function(data) {
+      if (data && data.length > 0 && data[0].Payload) {
+        let payload = JSON.parse(data[0].Payload);
+        
+        if (payload && payload.statusCode && payload.statusCode !== 200) {
+          console.log(\`Warm Up Invoke Success: \${functionName}\`, data);
+          success = false;
+          return;
+        }
+      }
+      
+      console.log(\`Warm Up Invoke Success: \${functionName}\`, data);
+    }, function(err) {
+      console.log(\`Warm Up Invoke Error: \${functionName}\`, err);
+      success = false;
+    });
+    return success;
```

Comment:
I think that the changes here are a step back in code quality.

It removes a lot of the functional/immutable code to replace with imperative/mutable code.

You could achieve the same with a bit better code. something like:
```javascript
    const warmUpFunction = `"use strict";

/** Generated by Serverless WarmUP Plugin at ${new Date().toISOString()} */
const aws = require("aws-sdk");
aws.config.region = "${this.options.region}";
const lambda = new aws.Lambda();
const functionDetails = ${JSON.stringify(functionDetails)};
module.exports.warmUp = async (event, context, callback) => {
  console.log("Warm Up Start");
  const invokes = await Promise.all(functionDetails.map(async ({ name, concurrency }) => {
    const params = {
      ClientContext: "${Buffer.from(`{"custom":${this.warmup.source}}`).toString('base64')}",
      FunctionName: name,
      InvocationType: "RequestResponse",
      LogType: "None",
      Qualifier: process.env.SERVERLESS_ALIAS || "$LATEST",
      Payload: '${this.warmup.source}'
    };

    try {
      await Promise.all(Array(concurrency).fill(0)
        .map(async _ => await lambda.invoke(params).promise()))
      console.log(\`Warm Up Invoke Success: \${name}\`);
      return true;
    } catch (e) {
      console.log(\`Warm Up Invoke Error: \${name}\`, e);
      return false;
    }
  }));

  console.log(\`Warm Up Finished with \${invokes.filter(r => !r).length} invoke errors\`);
}`
```

Reply:
Good idea, I'll update it.